src/librustc_mir/interpret/intern.rs

```diff
@@ -0,0 +1,219 @@
+//! This module specifies the type based interner for constants.
+//!
+//! After a const evaluation has computed a value, before we destroy the const evaluator's session
+//! memory, we need to extract all memory allocations to the global memory pool so they stay around.
+
+use rustc::ty::layout::LayoutOf;
+use rustc::ty;
+use rustc::mir::interpret::{
+    EvalResult, ErrorHandled,
+};
+use rustc::hir::{
+    self,
+    def_id::DefId,
+};
+use super::validity::RefTracking;
+use rustc_data_structures::fx::FxHashSet;
+
+use syntax::ast::Mutability;
+
+use super::{
+    ValueVisitor, MemoryKind, Pointer, AllocId, MPlaceTy, EvalErrorKind,
+};
+use crate::const_eval::{CompileTimeInterpreter, CompileTimeEvalContext};
+
+struct InternVisitor<'rt, 'a: 'rt, 'mir: 'rt, 'tcx: 'a+'rt+'mir> {
+    /// previously encountered safe references
+    ref_tracking: &'rt mut RefTracking<MPlaceTy<'tcx>>,
+    ecx: &'rt mut CompileTimeEvalContext<'a, 'mir, 'tcx>,
+    mode: InternMode,
+    /// A list of uninterned but seen relocations to be interned at the end if not interned by
+    /// a safe reference
+    leftover_relocations: &'rt mut FxHashSet<AllocId>,
```

Comment:
Actually we add all relocations there, right? Also I find the comment a bit hard to read.

What about

"A list of all encountered relocations. After type-based interning, we traverse this list to also intern allocations that are only referenced by a raw pointer or inside a union."